### PR TITLE
feat: new global / bound to grid local access schema

### PR DESCRIPTION
**Pull request by asalzburger**
Based on the discussion on #3012 - this PR introduces a different, harmonised way to do global -> grid_local or bound/local -> grid_local, more in line of what we do in the Magnetic field.

The owning delegate is needed for persistency and displaying.

If this is accepted as a new approach to do that, I will update also the grid based navigation structures with that.

### CODE REVIEW

**Comment by asalzburger** [comment]
Hi @dimitra97  - as you looked over the first PR #3012 (closed), this is the update that takes some discussion input into account. The idea is to have Global->GridLocal and Bound->GridLocal in delegates.

For the moment, this is only done for the `GridSurfaceMaterial` - but I plan to update the navigation delegates later to that as well (will also allow to implement a common IO for those).

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp [hunk 1]
```diff
@@ -92,52 +93,54 @@ struct IndexedMaterialAccessor {
...
-  /// @param transform  transform from global frame into map frame
+  /// @param boundToGridLocal the delegation from bound to grid local frame
+  /// @param globalToGridLocal the delegation from global into grid local frame
   GridSurfaceMaterialT(grid_type&& grid,
                        material_accessor_type&& materialAccessor,
-                       const std::vector<BinningValue>& gcasts,
-                       const std::vector<std::size_t>& laccessors,
-                       const Transform3& transform = Transform3::Identity())
+                       BoundToGridLocalDelegate&& boundToGridLocal,
+                       GlobalToGridLocalDelegate&& globalToGridLocal)
```
Comment (andiwand) [inline Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]:
```suggestion
                       BoundToGridLocalDelegate boundToGridLocal,
                       GlobalToGridLocalDelegate globalToGridLocal)
```

can this be passed by value so the user can decide if it should be moved or not?

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 2]
```diff
@@ -92,62 +93,164 @@ typename grid_type::point_t accessLocal(
...
+  GlobalSubspace() {
+    if constexpr (sizeof...(Args) == 0) {
+      throw std::invalid_argument(
+          "GlobalSubspace: cannot have an empty binning value list.");
+    }
+
+    if constexpr (sizeof...(Args) > 3) {
+      throw std::invalid_argument(
+          "GlobalSubspace: cannot have more than 3 binning values.");
+    }
```
Comment (andiwand) [inline Core/include/Acts/Utilities/GridAccessHelpers.hpp:159]: these should be static asserts outside the constructor which would cleanly separate runtime and compiletime errors

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 3]
```diff
@@ -92,62 +93,164 @@ typename grid_type::point_t accessLocal(
...
+// The bound to grid local transformation, if only access of a subspace
+// is requested
+template <std::size_t... Args>
+class LocalSubspace final : public IBoundToGridLocal {
+ public:
+  using grid_local_t = std::array<ActsScalar, sizeof...(Args)>;
+
+  // Constructor with sanity checks
+  LocalSubspace() {
+    if constexpr (sizeof...(Args) == 0) {
```
Comment (andiwand) [inline Core/include/Acts/Utilities/GridAccessHelpers.hpp:190]: same here

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]: Absolutely.

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp:159]: The point here is - that it should not be possible to event construct objects of this type ... hence, I would argue the code shouldn't compile ?!

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 4]
```diff
@@ -92,62 +93,164 @@ typename grid_type::point_t accessLocal(
...
+  static constexpr std::array<std::size_t, sizeof...(Args)> accessors = {
+      Args...};
+
+  /// Access the local entries
+  ///
+  /// @param lposition is the local position
+  ///
+  /// @return the grid position
+  grid_local_t toGridLocal(const Vector2& lposition) const {
+    // Fill the grid point from local according to the accessors
```
[the remark this answers is not in the document]
Comment (dimitra97) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp:228]: Yes ok the usage of Vector2 indicates that! Thanks!

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]: MMM, the OwningDelegate has (by design) a `unqiue_ptr` to the actual contained/owned class - hence, it looks like I am forced to move, isn't it?

Comment (andiwand) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]: Right but the move should be done by the caller I think. It is not strictly necessary to use an rvalue reference here. This is also what we do when we take an `unique_ptr` as an input.

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 5]
```diff
@@ -92,62 +93,164 @@ typename grid_type::point_t accessLocal(
...
+  /// @param position is the global position
+  ///
+  /// @return the grid position
+  typename global_to_grid_local_t::grid_local_t toGridLocal(
+      const Vector3& position) const {
+    return globalToGridLocal.toGridLocal(transform * position);
+  }
+};
+
+/// @brief A global (potentially castet) sub space of a global
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp:141]: Fixed.

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp]: Changed.

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp]: Changed.

**Comment by asalzburger** [comment]
Hey @andiwand @dimitra97  - I think I addressed your comments as much as I can.

File: Core/include/Acts/Material/GridSurfaceMaterial.hpp (hunk 1 above)
Comment (andiwand) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]: this does not work @asalzburger ?

Comment (asalzburger) [reply Core/include/Acts/Material/GridSurfaceMaterial.hpp:121]: Ahh, got it.

File: Core/include/Acts/Utilities/GridAccessHelpers.hpp [hunk 6]
```diff
@@ -92,62 +93,142 @@ typename grid_type::point_t accessLocal(
...
+class GlobalSubspace final : public IGlobalToGridLocal {
+ public:
+  using grid_local_t = std::array<ActsScalar, sizeof...(Args)>;
+
+  /// Assert that size has to be bigger than 0
+  static_assert(sizeof...(Args) > 0,
+                "GlobalSubspace: cannot have an empty binning value list.");
+
+  /// Asser that size has to be smaller than 4
+  static_assert(sizeof...(Args) <= 3,
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Utilities/GridAccessHelpers.hpp:154]: Fixed.

**Review by andiwand — APPROVED** [review]

**Review by dimitra97 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [30f97ea6b13dc2f54cbed592f5241e5a07f03f41]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37297206) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7091949)!

**Pull request merged**